crates/runtime/src/http/v1/mcp.rs

```diff
@@ -0,0 +1,140 @@
+    Extension,
+};
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{tools::mcp::server::RuntimeServer, Runtime};
+
+const FOUR_KB: usize = 1 << 12;
+
+type C2SWriter = Arc<Mutex<io::WriteHalf<io::SimplexStream>>>;
+type SessionId = Arc<str>;
+
+#[derive(Clone, Default)]
+pub struct McpState {
+    txs: Arc<tokio::sync::RwLock<HashMap<SessionId, C2SWriter>>>,
+}
+
+impl McpState {
+    pub(crate) async fn get(&self, session_id: &str) -> Option<C2SWriter> {
+        let rg = self.txs.read().await;
+        let writer = Arc::clone(rg.get(session_id)?);
+        Some(writer)
+    }
+}
+
+fn session_id() -> SessionId {
+    Arc::from(format!("{:016x}", rand::random::<u128>()))
+}
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct PostEventQuery {
+    pub session_id: String,
+}
+
+pub(crate) async fn sse(
+    Extension(rt): Extension<Arc<Runtime>>,
+    Extension(mcp): Extension<Arc<McpState>>,
+) -> Sse<impl Stream<Item = Result<Event, std::io::Error>>> {
+    let session = session_id();
+    tracing::trace!("New MCP connection with sessionid={session}");
+    let (c2s_read, c2s_write) = tokio::io::simplex(FOUR_KB);
+    let (s2c_read, s2c_write) = tokio::io::simplex(FOUR_KB);
+
+    mcp.txs
+        .write()
+        .await
+        .insert(Arc::clone(&session), Arc::new(Mutex::new(c2s_write)));
+    {
+        let session = Arc::clone(&session);
+        tokio::spawn(async move {
+            let server = Server::new(RouterService(RuntimeServer::from(&rt)));
+            let bytes_transport = ByteTransport::new(c2s_read, s2c_write);
+            let _result = server
+                .run(bytes_transport)
+                .await
+                .inspect_err(|e| tracing::error!(?e, "server run error"));
+            mcp.txs.write().await.remove(&session);
+        });
+    }
+
+    let stream = futures::stream::once(futures::future::ok(
+        Event::default()
+            .event("endpoint")
+            .data(format!("?sessionId={session}")),
+    ))
+    .chain(
+        FramedRead::new(
+            s2c_read,
+            crate::tools::mcp::server::codec::JsonRpcFrameCodec,
+        )
+        .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+        .and_then(move |bytes| match std::str::from_utf8(&bytes) {
+            Ok(message) => futures::future::ok(Event::default().event("message").data(message)),
+            Err(e) => futures::future::err(io::Error::new(io::ErrorKind::InvalidData, e)),
+        }),
+    );
+    Sse::new(stream)
+}
+
+pub(crate) async fn event(
+    Extension(mcp): Extension<Arc<McpState>>,
+    Query(PostEventQuery { session_id }): Query<PostEventQuery>,
+    body: Bytes,
+) -> Result<StatusCode, StatusCode> {
+    const BODY_BYTES_LIMIT: usize = 1 << 22;
+    tracing::trace!(
+        "Received POST event in SSE session_id={session_id}. Event={}",
+        String::from_utf8(body.to_ascii_lowercase()).unwrap_or("ERROR".to_string())
```

Comment:
```suggestion
        String::from_utf8_lossy(body)
```

converting the entire body to lowercase could potentially be expensive.

Reply:
Will do in follow up